time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalDateTimeArbitrary.java

```diff
@@ -0,0 +1,151 @@
+package net.jqwik.time.internal.properties.arbitraries;
+
+import java.time.*;
+import java.time.temporal.*;
+
+import org.apiguardian.api.*;
+
+import net.jqwik.api.*;
+import net.jqwik.api.arbitraries.*;
+import net.jqwik.time.api.*;
+import net.jqwik.time.api.arbitraries.*;
+
+import static java.time.temporal.ChronoUnit.*;
+import static org.apiguardian.api.API.Status.*;
+
+@API(status = INTERNAL)
+public class DefaultLocalDateTimeArbitrary extends ArbitraryDecorator<LocalDateTime> implements LocalDateTimeArbitrary {
+
+	private static final LocalDateTime DEFAULT_MIN = LocalDateTime.of(DefaultLocalDateArbitrary.DEFAULT_MIN_DATE, LocalTime.MIN);
+	private static final LocalDateTime DEFAULT_MAX = LocalDateTime.of(DefaultLocalDateArbitrary.DEFAULT_MAX_DATE, LocalTime.MAX);
+
+	private LocalDateTime min = null;
+	private LocalDateTime max = null;
+
+	private ChronoUnit ofPrecision = DefaultLocalTimeArbitrary.DEFAULT_PRECISION;
+
+	@Override
+	protected Arbitrary<LocalDateTime> arbitrary() {
+
+		LocalDateTime effectiveMin = calculateEffectiveMin();
+		LocalDateTime effectiveMax = calculateEffectiveMax();
+		if (effectiveMax.isBefore(effectiveMin)) {
+			throw new IllegalArgumentException("The maximum date time is too soon after the minimum date time.");
+		}
+
+		LocalDateArbitrary dates = Dates.dates();
+		LocalTimeArbitrary[] times = generateTimeArbitrary(effectiveMin, effectiveMax, ofPrecision);
+
+		dates = dates.atTheEarliest(effectiveMin.toLocalDate());
+		dates = dates.atTheLatest(effectiveMax.toLocalDate());
+
+		LocalDate effectiveMinDate = effectiveMin.toLocalDate();
+		LocalDate effectiveMaxDate = effectiveMax.toLocalDate();
+
+		return dates.flatMap(date -> timesByDate(times, date, effectiveMinDate, effectiveMaxDate)
+											 .map(time -> LocalDateTime.of(date, time)));
+
+	}
+
+	private LocalTimeArbitrary timesByDate(LocalTimeArbitrary[] times, LocalDate date, LocalDate effectiveMin, LocalDate effectiveMax) {
+		if (date.isEqual(effectiveMin)) {
+			return times[0];
+		} else if (date.isEqual(effectiveMax)) {
+			return times[2];
+		} else {
+			return times[1];
+		}
+	}
+
+	private LocalDateTime calculateEffectiveMin() {
+		LocalDateTime effective = min != null ? min : DEFAULT_MIN;
+		return calculateEffectiveMinWithPrecision(effective);
+	}
+
+	private LocalDateTime calculateEffectiveMinWithPrecision(LocalDateTime effective) {
+		LocalDate date = effective.toLocalDate();
+		LocalTime time = effective.toLocalTime();
+		try {
+			time = DefaultLocalTimeArbitrary.calculateEffectiveMinWithPrecision(time, ofPrecision);
+		} catch (IllegalArgumentException e) {
+			time = LocalTime.MIN;
+			LocalDate effectiveDate;
+			try {
+				effectiveDate = date.plusDays(1);
+			} catch (DateTimeException dateTimeException) {
+				throw new IllegalArgumentException("Date is LocalDate.MAX and must be increased by 1 day.");
```

Comment:
In what way is this referring to an illegal argument and how does the error message help the user to figure out what's wrong?

Reply:
I'll change it to:
`throw new IllegalArgumentException("Min value must be increased but results in a " + e.getMessage());`
like in `DefaultDurationArbitrary`

Reply:
It must be increased because of the ofPrecision value. Is this better?:
`throw new IllegalArgumentException("Min value must be increased because of precision but results in a " + e.getMessage());`

Reply:
Example:
`min: 23:59:32, ofPrecision: SECONDS`
=> `effectiveMin: 00:00:00` => overflow

Reply:
Now I understand (I hope). I think I’d prefer an error when trying to set a min/max date(time) which has more precision than the chosen precision. Or maybe increase precision implicitly if value requires it and it’s not explicitly specified.

Reply:
But your example doesn’t make sense because SECONDS should allow seconds but you seem to round up to minutes.

Reply:
In my opinion an IllegalArgumentException at this position in the code is the best way.

Problem when this exception is thrown when the user sets the values:
`Times.times().atTheEarliest(LocalTime.of(23, 59, 59, 921_000_000)).ofPrecision(MILLIS)`
throws an exception, but is a valid statement

Implicitly increase the precision is not what I expect as a user. And there is a other problem:
`Times.times().atTheEarliest(LocalTime.of(23, 59, 59, 921_000_000)).ofPrecision(SECONDS)`
has to thrown an exception. So the exception is not gone.

So I think the current way is the best way. What do you think?

Reply:
Hm. When I use a value with odd millis why would I NOT expect this to imply a precision? And yes the last example would throw an exception which can tell me what’s wrong: „Cannot use milliseconds with precision SECONDS“ would perfectly explain what’s wrong.

Reply:
My feeling said that.

So you think I should change it to change the precision implicitly?

Reply:
I'd vote for yes ;-)

Reply:
`But your example doesn’t make sense because SECONDS should allow seconds but you seem to round up to minutes.`

It didn't show me that comment yesterday. I saw and corrected it shortly after posting the example.

Reply:
Done.